core/src/test/java/org/infinispan/statetransfer/ConcurrentStartChanelLookupTest.java

```diff
@@ -0,0 +1,118 @@
+package org.infinispan.statetransfer;
+
+import org.infinispan.Cache;
+import org.infinispan.configuration.cache.CacheMode;
+import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.configuration.global.GlobalConfigurationBuilder;
+import org.infinispan.lifecycle.ComponentStatus;
+import org.infinispan.manager.DefaultCacheManager;
+import org.infinispan.manager.EmbeddedCacheManager;
+import org.infinispan.test.MultipleCacheManagersTest;
+import org.infinispan.test.fwk.CleanupAfterMethod;
+import org.infinispan.test.fwk.JGroupsConfigBuilder;
+import org.infinispan.test.fwk.TestResourceTracker;
+import org.infinispan.test.fwk.TransportFlags;
+import org.jgroups.JChannel;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.util.concurrent.Future;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.infinispan.test.TestingUtil.blockUntilViewsReceived;
+import static org.infinispan.test.TestingUtil.extractGlobalComponentRegistry;
+import static org.infinispan.test.TestingUtil.waitForRehashToComplete;
+import static org.testng.AssertJUnit.assertEquals;
+
+/**
+ * Tests concurrent startup of replicated and distributed caches
+ *
+ * @author Dan Berindei
+ * @since 7.2
+ */
+@Test(testName = "statetransfer.ConcurrentStartTest", groups = "functional")
+@CleanupAfterMethod
+public class ConcurrentStartChanelLookupTest extends MultipleCacheManagersTest {
+
+   @Override
+   protected void createCacheManagers() throws Throwable {
+      // The test method will create the cache managers
+   }
+
+   @DataProvider(name = "startOrder")
+   public Object[][] startOrder() {
+      return new Object[][]{{0, 1}, {1, 0}};
+   }
+
+   @Test(timeOut = 60000, dataProvider = "startOrder")
+   public void testConcurrentStart(int eagerManager, int lazyManager) throws Exception {
+      TestResourceTracker.testThreadStarted(this);
+      String name1 = TestResourceTracker.getNextNodeName();
+      String name2 = TestResourceTracker.getNextNodeName();
+
+      // Create and connect both channels beforehand
+      // We need both nodes in the view when the coordinator's ClusterTopologyManagerImpl starts
+      // in order to reproduce the ISPN-5106 deadlock
+      JChannel ch1 = createChannel(name1, 0);
+      JChannel ch2 = createChannel(name2, 1);
+
+      // Use a JGroupsChannelLookup to pass the created channels to the transport
+      EmbeddedCacheManager cm1 = createCacheManager(name1, ch1);
+      EmbeddedCacheManager cm2 = createCacheManager(name2, ch2);
+
+      try {
+         assertEquals(ComponentStatus.INSTANTIATED, extractGlobalComponentRegistry(cm1).getStatus());
+         assertEquals(ComponentStatus.INSTANTIATED, extractGlobalComponentRegistry(cm2).getStatus());
+
+         log.debugf("Channels created. Starting the caches");
+         Future<Object> repl1Future = fork(() -> manager(eagerManager).getCache("repl"));
+
+         Thread.sleep(1000);
```

Comment:
It's tricky, because we don't know exactly when JGroups sends the message, and in some cases it can even retransmit messages that were sent right before the join. But I could reduce the timeout :)

There is another option, we could emulate the lost commands by intercepting incoming commands on the second node and discarding the commands there. Still, I wanted this test to "document" how injected channels behave.